src/memory_manager.rs

```diff
@@ -745,6 +745,45 @@ pub fn add_finalizer<VM: VMBinding>(
     mmtk.finalizable_processor.lock().unwrap().add(object);
 }
 
+/// Pin an object. MMTk will make sure that the object does not move
+/// during GC. Note that action cannot happen in some plans, eg, semispace.
+/// It returns true if the pinning operation has been performed, i.e.,
```

Comment:
I have the same question here as Wenyu. Intuitively, for this method, , we could return either:
1. if this call to `pin_object` successfully pinned the object (i.e. is the action successful?), or
2. if the object is pinned after this call (i.e. is the state as expected?).

Is there any reason that you chose 1 rather than 2?

Reply:
For example, if a binding both use Immix (moving) and mark sweep. The binding's code that calls `pin_object` should be agnostic to the actual plan used. For Immix, it calls `pin_object`, and expects it to return `true` so the binding knows the object will not be moved. But for mark sweep, it also calls `pin_object`, and MMTk returns `false` even if the object will not be moved. That could cause some confusion.

Reply:
Sometimes you want to know whether you actually did the pinning so you can safely unpin later. Say for instance an object has been pinned previously, and you have some code that does:
```
pin_object(a)
// do GC
unpin_object(a)
// do GC
``` 
Now `a` can move, even though it was pinned somewhere else and it was not supposed to move. Whereas you may have:
```
let pinned = pin_object(a)
// do GC
if pinned
  unpin_object(a)
// do GC
``` 
One option would be keeping count of pins and unpins, but that may be too expensive, at least for now (?)

Reply:
Furthermore `pin_object` guarantees that the object should not move (unless it is movable and it is unpinned later), even when it returns `false`. In that sense, the 2nd piece of code from the previous comment would work for both mark sweep and immix.

Reply:
> Furthermore `pin_object` guarantees that the object should not move (unless it is movable and it is unpinned later), even when it returns `false`. In that sense, the 2nd piece of code from the previous comment would work for both mark sweep and immix.

Yeah, this sounds reasonable.